// feishu-agent-bridge/dist/index.js
// src/index.ts
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import { createRequire } from "module";
import * as Lark2 from "@larksuiteoapi/node-sdk";
import { z as z2 } from "zod";

// src/types.ts
import { z } from "zod";
var AutoPromptSchema = z.object({
  enabled: z.boolean().default(false),
  intervalSeconds: z.number().int().positive().max(300).default(30),
  maxIterations: z.number().int().positive().max(100).default(10),
  message: z.string().min(1).default("\u8BF7\u540C\u6B65\u5F53\u524D\u8FDB\u5EA6\uFF0C\u5982\u9700\u5E2E\u52A9\u8BF7\u8BF4\u660E")
});
var FeishuConfigSchema = z.object({
  appId: z.string().min(1, "appId \u4E0D\u80FD\u4E3A\u7A7A"),
  appSecret: z.string().min(1, "appSecret \u4E0D\u80FD\u4E3A\u7A7A"),
  timeout: z.number().int().positive().max(6e5).default(12e4),
  logLevel: z.enum(["fatal", "error", "warn", "info", "debug", "trace"]).default("info"),
  maxHistoryMessages: z.number().int().positive().max(500).default(200),
  pollInterval: z.number().int().positive().default(1e3),
  stablePolls: z.number().int().positive().default(3),
  dedupTtl: z.number().int().positive().default(10 * 60 * 1e3),
  directory: z.string().optional()
});

// src/feishu/gateway.ts
async function startFeishuGateway(options) {
  const { log } = options;
  log("info", "\u98DE\u4E66 HTTP \u6A21\u5F0F\uFF1A\u7531\u5BBF\u4E3B\u63D0\u4F9B Webhook\uFF08\u5982 POST /webhook/feishu\uFF09");
  return {
    shutdown: async () => {
      log("info", "\u5173\u95ED\u98DE\u4E66 HTTP \u5360\u4F4D\u7F51\u5173");
    }
  };
}

// src/feishu/gateway-ws.ts
import * as Lark from "@larksuiteoapi/node-sdk";
import { HttpProxyAgent } from "http-proxy-agent";
import { HttpsProxyAgent } from "https-proxy-agent";

// src/feishu/dedup.ts
var seen = /* @__PURE__ */ new Map();
var DEFAULT_TTL_MS = 2 * 60 * 1e3;
function prune(now, ttl) {
  for (const [id, t] of seen) {
    if (now - t > ttl) seen.delete(id);
  }
}
function isDuplicateMessageId(messageId, ttlMs = DEFAULT_TTL_MS) {
  if (!messageId) return false;
  const now = Date.now();
  prune(now, ttlMs);
  if (seen.has(messageId)) return true;
  seen.set(messageId, now);
  return false;
}

// src/feishu/gateway-ws.ts
function extractTextContent(message) {
  try {
    if (message.message_type === "text") {
      const content = JSON.parse(message.content || "{}");
      return content.text || "";
    }
  } catch {
  }
  return "";
}
function computeShouldReply(chatType, message, botOpenId) {
  if (chatType === "p2p") return true;
  const mentions = message.mentions;
  if (Array.isArray(mentions) && mentions.length > 0) {
    return mentions.some((m) => m?.id?.open_id === botOpenId);
  }
  if (message.message_type === "text") {
    try {
      const content = JSON.parse(message.content || "{}");
      const text = content.text || "";
      return text.includes(`<at user_id="${botOpenId}">`) || text.includes(`<at open_id="${botOpenId}">`);
    } catch {
      return false;
    }
  }
  return false;
}
function startFeishuWebSocketGateway(options) {
  const { config, botOpenId, handlers, log, httpInstance } = options;
  const { appId, appSecret } = config;
  const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY || process.env.ALL_PROXY || "";
  let wsAgent;
  if (proxyUrl) {
    wsAgent = new HttpsProxyAgent(proxyUrl);
    log("info", "\u98DE\u4E66 WS \u5DF2\u542F\u7528\u4EE3\u7406", { proxy: proxyUrl });
    if (!globalThis.httpsAgent) {
      ;
      globalThis.httpsAgent = wsAgent;
    }
    if (proxyUrl.startsWith("http://")) {
      const httpAgent = new HttpProxyAgent(proxyUrl);
      if (!globalThis.httpAgent) {
        ;
        globalThis.httpAgent = httpAgent;
      }
      log("info", "\u98DE\u4E66 WS HTTP Agent \u5DF2\u914D\u7F6E", { proxy: proxyUrl });
    }
  }
  const dispatcher = new Lark.EventDispatcher({}).register({
    "im.message.receive_v1": async (data) => {
      try {
        const message = data.message;
        if (!message) return;
        const chatId = message.chat_id;
        if (!chatId) return;
        const messageId = message.message_id;
        if (isDuplicateMessageId(messageId)) return;
        const messageType = message.message_type || "text";
        let text = extractTextContent(message);
        if (messageType === "text") {
          text = text.replace(/@_user_\d+\s*/g, "").trim();
        }
        if (!text) return;
        const chatType = message.chat_type === "group" ? "group" : "p2p";
        const shouldReply = computeShouldReply(chatType, message, botOpenId);
        const sender = data.sender;
        const senderId = sender?.sender_id?.open_id || sender?.sender_id?.user_id || "";
        const ctx = {
          chatId,
          messageId: messageId || "",
          messageType,
          content: text,
          rawContent: message.content || "{}",
          chatType,
          senderId,
          rootId: message.root_id,
          createTime: message.create_time,
          shouldReply
        };
        log("info", "WS \u6536\u5230\u98DE\u4E66\u6D88\u606F", {
          chatId,
          messageId: messageId || "",
          chatType,
          shouldReply,
          preview: text.slice(0, 80)
        });
        if (handlers.onMessage) {
          await handlers.onMessage(ctx);
        }
      } catch (err) {
        log("error", "WS \u5904\u7406\u6D88\u606F\u5931\u8D25", {
          error: err instanceof Error ? err.message : String(err)
        });
      }
    },
    "im.chat.member.bot.added_v1": async (data) => {
      try {
        const chatId = data.chat_id || data.detail?.event?.chat_id;
        if (chatId && handlers.onBotAdded) {
          log("info", "WS: Bot \u5165\u7FA4", { chatId });
          await handlers.onBotAdded(chatId);
        }
      } catch (err) {
        log("error", "WS \u5904\u7406\u5165\u7FA4\u5931\u8D25", {
          error: err instanceof Error ? err.message : String(err)
        });
      }
    }
  });
  const logLevelMap = {
    fatal: Lark.LoggerLevel.fatal,
    error: Lark.LoggerLevel.error,
    warn: Lark.LoggerLevel.warn,
    info: Lark.LoggerLevel.info,
    debug: Lark.LoggerLevel.debug,
    trace: Lark.LoggerLevel.trace
  };
  const wsClient = new Lark.WSClient({
    appId,
    appSecret,
    domain: Lark.Domain.Feishu,
    ...httpInstance ? { httpInstance } : {},
    ...wsAgent ? { agent: wsAgent } : {},
    loggerLevel: logLevelMap[config.logLevel] ?? Lark.LoggerLevel.info,
    logger: {
      error: (...msg) => log("error", "[lark.ws]", { msg }),
      warn: (...msg) => log("warn", "[lark.ws]", { msg }),
      info: (...msg) => log("info", "[lark.ws]", { msg }),
      debug: (...msg) => log("info", "[lark.ws]", { msg }),
      trace: (...msg) => log("info", "[lark.ws]", { msg })
    }
  });
  wsClient.start({ eventDispatcher: dispatcher });
  log("info", "\u98DE\u4E66 WebSocket \u957F\u8FDE\u63A5\u5DF2\u542F\u52A8", { appIdPrefix: appId.slice(0, 8) + "..." });
  log("info", "\u98DE\u4E66\u4FA7\u8BF7\u9009\u62E9\u300C\u4F7F\u7528\u957F\u8FDE\u63A5\u63A5\u6536\u4E8B\u4EF6\u300D\uFF0C\u65E0\u9700\u914D\u7F6E\u8BF7\u6C42\u5730\u5740");
  return {
    shutdown: async () => {
      log("info", "\u98DE\u4E66 WebSocket \u6B63\u5728\u5173\u95ED");
      wsClient.close();
      log("info", "\u98DE\u4E66 WebSocket \u5DF2\u5173\u95ED");
    }
  };
}

// src/feishu/sender.ts
var FeishuSender = class {
  constructor(larkClient, log) {
    this.larkClient = larkClient;
    this.log = log;
  }
  /**
   * 发送文本消息
   */
  async sendText(chatId, text) {
    try {
      const res = await this.larkClient.im.message.create({
        params: {
          receive_id_type: "chat_id"
        },
        data: {
          receive_id: chatId,
          msg_type: "text",
          content: JSON.stringify({ text })
        }
      });
      const messageId = res?.data?.message_id;
      if (messageId) {
        this.log("info", "\u6D88\u606F\u5DF2\u53D1\u9001", { chatId, messageId });
        return true;
      } else {
        this.log("error", "\u53D1\u9001\u6D88\u606F\u5931\u8D25", { chatId, response: res });
        return false;
      }
    } catch (err) {
      this.log("error", "\u53D1\u9001\u6D88\u606F\u5F02\u5E38", {
        chatId,
        error: err instanceof Error ? err.message : String(err)
      });
      return false;
    }
  }
  /**
   * 发送富文本卡片
   */
  async sendCard(chatId, card) {
    try {
      const res = await this.larkClient.im.message.create({
        params: {
          receive_id_type: "chat_id"
        },
        data: {
          receive_id: chatId,
          msg_type: "interactive",
          card
        }
      });
      const messageId = res?.data?.message_id;
      if (messageId) {
        this.log("info", "\u5361\u7247\u5DF2\u53D1\u9001", { chatId, messageId });
        return true;
      } else {
        this.log("error", "\u53D1\u9001\u5361\u7247\u5931\u8D25", { chatId, response: res });
        return false;
      }
    } catch (err) {
      this.log("error", "\u53D1\u9001\u5361\u7247\u5F02\u5E38", {
        chatId,
        error: err instanceof Error ? err.message : String(err)
      });
      return false;
    }
  }
  /**
   * 更新消息（用于流式响应）
   */
  async updateMessage(messageId, text) {
    try {
      await this.larkClient.im.message.patch({
        path: {
          message_id: messageId
        },
        data: {
          content: JSON.stringify({ text })
        }
      });
      this.log("info", "\u6D88\u606F\u5DF2\u66F4\u65B0", { messageId });
      return true;
    } catch (err) {
      this.log("error", "\u66F4\u65B0\u6D88\u606F\u5F02\u5E38", {
        messageId,
        error: err instanceof Error ? err.message : String(err)
      });
      return false;
    }
  }
};
function createSender(larkClient, log) {
  return new FeishuSender(larkClient, log);
}

// src/index.ts
var require2 = createRequire(import.meta.url);
var SERVICE_NAME = "feishu-agent-bridge";
var LOG_PREFIX = "[feishu]";
var isDebug = !!process.env.FEISHU_DEBUG;
function createProxyAwareHttpInstance() {
  const httpsProxy = process.env.HTTPS_PROXY || process.env.https_proxy;
  if (!httpsProxy) {
    return Lark2.defaultHttpInstance;
  }
  let httpsAgent = void 0;
  try {
    if (httpsProxy) {
      const { HttpsProxyAgent: HttpsProxyAgentClass } = require2("https-proxy-agent");
      httpsAgent = new HttpsProxyAgentClass(httpsProxy);
    }
  } catch (err) {
    console.warn(`[feishu] \u65E0\u6CD5\u52A0\u8F7D https-proxy-agent: ${err.message}`);
  }
  if (httpsAgent) {
    console.log(`[feishu] \u914D\u7F6E HTTPS \u4EE3\u7406: ${httpsProxy}`);
  }
  return {
    request: async (opts) => {
      const config = { ...opts, httpAgent: httpsAgent, httpsAgent };
      return Lark2.defaultHttpInstance.request(config);
    },
    get: async (url, opts) => Lark2.defaultHttpInstance.get(url, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    post: async (url, data, opts) => Lark2.defaultHttpInstance.post(url, data, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    put: async (url, data, opts) => Lark2.defaultHttpInstance.put(url, data, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    patch: async (url, data, opts) => Lark2.defaultHttpInstance.patch(url, data, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    delete: async (url, opts) => Lark2.defaultHttpInstance.delete(url, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    head: async (url, opts) => Lark2.defaultHttpInstance.head(url, { ...opts, httpAgent: httpsAgent, httpsAgent }),
    options: async (url, opts) => Lark2.defaultHttpInstance.options(url, { ...opts, httpAgent: httpsAgent, httpsAgent })
  };
}
function parseTransport(raw) {
  const v = (raw || "http").toLowerCase().trim();
  if (v === "ws" || v === "websocket" || v === "long") return "ws";
  if (v === "both" || v === "dual") return "both";
  return "http";
}
var DEFAULT_FEISHU_CONFIG_PATHS = [
  () => join(homedir(), ".config", "feishu-agent-bridge", "feishu.json"),
  () => join(homedir(), ".config", "cursor", "plugins", "feishu.json")
];
async function createFeishuService(options) {
  const {
    config: partialConfig,
    appId: optAppId,
    appSecret: optAppSecret,
    transport: transportOpt,
    onMessage,
    onBotAdded,
    onCardAction,
    log: customLog
  } = options;
  const effectivePartial = partialConfig ?? (optAppId && optAppSecret ? { appId: optAppId, appSecret: optAppSecret } : void 0);
  const log = customLog || ((level, message, extra) => {
    const prefixed = `${LOG_PREFIX} ${message}`;
    if (isDebug) {
      console.error(JSON.stringify({
        ts: (/* @__PURE__ */ new Date()).toISOString(),
        service: SERVICE_NAME,
        level,
        message: prefixed,
        ...extra
      }));
    } else if (level === "error" || level === "warn") {
      console.error(prefixed, extra || "");
    } else {
      console.log(prefixed, extra || "");
    }
  });
  const config = await loadConfig(effectivePartial, log);
  const transport = transportOpt ?? parseTransport(process.env.FEISHU_TRANSPORT);
  const httpInstance = createProxyAwareHttpInstance();
  const larkClient = new Lark2.Client({
    appId: config.appId,
    appSecret: config.appSecret,
    domain: Lark2.Domain.Feishu,
    appType: Lark2.AppType.SelfBuild,
    httpInstance
  });
  const botOpenId = await fetchBotOpenId(larkClient, log);
  const sender = createSender(larkClient, log);
  const handlers = {
    onMessage,
    onBotAdded,
    onCardAction
  };
  let wsGateway = null;
  let httpGateway = null;
  if (transport === "ws" || transport === "both") {
    wsGateway = startFeishuWebSocketGateway({
      config,
      larkClient,
      httpInstance,
      botOpenId,
      handlers,
      log
    });
  }
  log("info", "\u98DE\u4E66\u670D\u52A1\u5DF2\u521D\u59CB\u5316", {
    appId: config.appId.slice(0, 8) + "...",
    botOpenId,
    transport
  });
  return {
    transport,
    async run() {
      try {
        if (transport === "http" || transport === "both") {
          httpGateway = await startFeishuGateway({
            config,
            larkClient,
            botOpenId,
            handlers,
            log
          });
        }
        log("info", "\u98DE\u4E66\u670D\u52A1\u5DF2\u8FDB\u5165\u8FD0\u884C\u6001", { transport });
        await new Promise(() => {
        });
      } catch (err) {
        log("error", "\u542F\u52A8\u98DE\u4E66\u670D\u52A1\u5931\u8D25", {
          error: err instanceof Error ? err.message : String(err)
        });
        throw err;
      }
    },
    async shutdown() {
      if (wsGateway) {
        await wsGateway.shutdown();
        wsGateway = null;
      }
      if (httpGateway) {
        await httpGateway.shutdown();
        httpGateway = null;
      }
      log("info", "\u98DE\u4E66\u670D\u52A1\u5DF2\u5173\u95ED");
    },
    getSender() {
      return sender;
    },
    getClient() {
      return larkClient;
    }
  };
}
async function loadConfig(partialConfig, log) {
  if (partialConfig && partialConfig.appId && partialConfig.appSecret) {
    const config = FeishuConfigSchema.parse({
      appId: partialConfig.appId,
      appSecret: partialConfig.appSecret,
      ...partialConfig
    });
    return {
      ...config,
      directory: expandDirectoryPath(partialConfig.directory || "")
    };
  }
  let configPath = null;
  for (const getPath of DEFAULT_FEISHU_CONFIG_PATHS) {
    const p = getPath();
    if (existsSync(p)) {
      configPath = p;
      break;
    }
  }
  if (!configPath) {
    const primary = DEFAULT_FEISHU_CONFIG_PATHS[0]();
    const legacy = DEFAULT_FEISHU_CONFIG_PATHS[1]();
    throw new Error(
      `\u7F3A\u5C11\u98DE\u4E66\u914D\u7F6E\u6587\u4EF6\uFF1A\u8BF7\u521B\u5EFA ${primary}\uFF08\u63A8\u8350\uFF09\uFF0C\u6216\u6CBF\u7528\u517C\u5BB9\u8DEF\u5F84 ${legacy}\uFF0C\u5185\u5BB9\u4E3A {"appId":"cli_xxx","appSecret":"xxx"}\uFF1B\u4E5F\u53EF\u5728\u4EE3\u7801\u4E2D\u4F20\u5165 appId / appSecret \u6216\u5B8C\u6574 config`
    );
  }
  try {
    const raw = resolveEnvPlaceholders(JSON.parse(readFileSync(configPath, "utf-8")));
    const config = FeishuConfigSchema.parse(raw);
    return {
      ...config,
      directory: expandDirectoryPath(config.directory || "")
    };
  } catch (err) {
    if (err instanceof z2.ZodError) {
      const details = err.issues.map((i) => `  - ${i.path.join(".")}: ${i.message}`).join("\n");
      throw new Error(`${LOG_PREFIX} \u914D\u7F6E\u9A8C\u8BC1\u5931\u8D25:
${details}`);
    }
    if (err instanceof SyntaxError) {
      throw new Error(`\u98DE\u4E66\u914D\u7F6E\u6587\u4EF6\u683C\u5F0F\u9519\u8BEF\uFF1A${configPath} \u5FC5\u987B\u662F\u5408\u6CD5\u7684 JSON (${err.message})`);
    }
    throw err;
  }
}
function expandDirectoryPath(dir) {
  if (!dir) return dir;
  if (dir.startsWith("~")) {
    dir = join(homedir(), dir.slice(1));
  }
  dir = dir.replace(/\$\{(\w+)\}/g, (_match, name) => {
    const val = process.env[name];
    if (val === void 0) {
      throw new Error(`\u73AF\u5883\u53D8\u91CF ${name} \u672A\u8BBE\u7F6E\uFF08directory \u5F15\u7528\u4E86 \${${name}}\uFF09`);
    }
    return val;
  });
  return dir;
}
function resolveEnvPlaceholders(obj) {
  if (typeof obj === "string") {
    if (!obj.includes("${")) return obj;
    return obj.replace(/\$\{(\w+)\}/g, (_match, name) => {
      const val = process.env[name];
      if (val === void 0) {
        throw new Error(`\u73AF\u5883\u53D8\u91CF ${name} \u672A\u8BBE\u7F6E\uFF08\u914D\u7F6E\u503C\u5F15\u7528\u4E86 \${${name}}\uFF09`);
      }
      return val;
    });
  }
  if (Array.isArray(obj)) {
    return obj.map(resolveEnvPlaceholders);
  }
  if (obj !== null && typeof obj === "object") {
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
      result[key] = resolveEnvPlaceholders(value);
    }
    return result;
  }
  return obj;
}
async function fetchBotOpenId(larkClient, log) {
  const res = await larkClient.request({
    url: "https://open.feishu.cn/open-apis/bot/v3/info",
    method: "GET"
  });
  const openId = res?.bot?.open_id;
  if (!openId) {
    throw new Error("Bot open_id \u4E3A\u7A7A\uFF0C\u65E0\u6CD5\u542F\u52A8\u98DE\u4E66\u670D\u52A1");
  }
  log("info", "Bot open_id \u83B7\u53D6\u6210\u529F", { openId });
  return openId;
}
export {
  DEFAULT_FEISHU_CONFIG_PATHS,
  createFeishuService,
  createSender,
  startFeishuGateway,
  startFeishuWebSocketGateway
};
